src/Microsoft.ML.TimeSeries/RootCauseLocalizationType.cs

```diff
@@ -0,0 +1,308 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.ML.Data;
+
+namespace Microsoft.ML.TimeSeries
+{
+    /// <summary>
+    /// Allows a member to be marked as a <see cref="RootCauseLocalizationInputDataViewType"/>, primarily allowing one to set
+    /// root cause localization input.
+    /// </summary>
+    public sealed class RootCauseLocalizationInputTypeAttribute : DataViewTypeAttribute
+    {
+        /// <summary>
+        /// Create a root cause localizagin input type.
+        /// </summary>
+        public RootCauseLocalizationInputTypeAttribute()
+        {
+        }
+
+        /// <summary>
+        /// Equal function.
+        /// </summary>
+        public override bool Equals(DataViewTypeAttribute other)
+        {
+            if (!(other is RootCauseLocalizationInputTypeAttribute otherAttribute))
+                return false;
+            return true;
+        }
+
+        /// <summary>
+        /// Produce the same hash code for all RootCauseLocalizationInputTypeAttribute.
+        /// </summary>
+        public override int GetHashCode()
+        {
+            return 0;
+        }
+
+        public override void Register()
+        {
+            DataViewTypeManager.Register(new RootCauseLocalizationInputDataViewType(), typeof(RootCauseLocalizationInput), this);
+        }
+    }
+
+    /// <summary>
+    /// Allows a member to be marked as a <see cref="RootCauseDataViewType"/>, primarily allowing one to set
+    /// root cause result.
+    /// </summary>
+    public sealed class RootCauseTypeAttribute : DataViewTypeAttribute
+    {
+        /// <summary>
+        /// Create an root cause type.
+        /// </summary>
+        public RootCauseTypeAttribute()
+        {
+        }
+
+        /// <summary>
+        /// RootCauseTypeAttribute with the same type should equal.
+        /// </summary>
+        public override bool Equals(DataViewTypeAttribute other)
+        {
+            if (other is RootCauseTypeAttribute otherAttribute)
+                return true;
+            return false;
+        }
+
+        /// <summary>
+        /// Produce the same hash code for all RootCauseTypeAttribute.
+        /// </summary>
+        public override int GetHashCode()
+        {
+            return 0;
+        }
+
+        public override void Register()
+        {
+            DataViewTypeManager.Register(new RootCauseDataViewType(), typeof(RootCause), this);
+        }
+    }
+
+    public sealed class RootCause
+    {
+        public List<RootCauseItem> Items { get; set; }
+        public RootCause()
+        {
+            Items = new List<RootCauseItem>();
+        }
+    }
+
+    public sealed class RootCauseLocalizationInput
+    {
```

Comment:
This is a fairly complex class to be treated as the input column. 
One common scenario would be that I have a file with data of primitive types and timestamp columns. I want to detect anomalies on a particular column and then detect the root causes that led to those anomalies. 

Are we expecting the users to take the data from the text loader and perform the conversion to RootCauseLocalizationInput? 
Is it possible to do this on behalf of the user transparently?

Reply:
Yes, according to the RCA request, the input and output is much more complex compared to anomaly detection.  According to current ML.net design,  maybe your requirement is hard to implement. We have discussed in this problem in email, I will forward the email to you then.

Reply:
> This is a fairly complex class to be treated as the input column.
> One common scenario would be that I have a file with data of primitive types and timestamp columns. I want to detect anomalies on a particular column and then detect the root causes that led to those anomalies.
> 
> Are we expecting the users to take the data from the text loader and perform the conversion to RootCauseLocalizationInput?
> Is it possible to do this on behalf of the user transparently?

Yes, according to the RCA request, the input and output is much more complex compared to anomaly detection. According to current ML.net design, maybe your requirement is hard to implement. We have discussed in this problem in email, I will forward the email to you then.

Reply:
> > This is a fairly complex class to be treated as the input column.
> > One common scenario would be that I have a file with data of primitive types and timestamp columns. I want to detect anomalies on a particular column and then detect the root causes that led to those anomalies.
> > Are we expecting the users to take the data from the text loader and perform the conversion to RootCauseLocalizationInput?
> > Is it possible to do this on behalf of the user transparently?
> 
> Yes, according to the RCA request, the input and output is much more complex compared to anomaly detection. According to current ML.net design, maybe your requirement is hard to implement. We have discussed in this problem in email, I will forward the email to you then.

For the user scenario youo mentioned, both SRCNN and Root cause are libary that users can use, When they have the data, the can use this lib in their way